File: compiler/src/main/java/com/bennyhuo/compiler/OptionalField.java
package com.bennyhuo.compiler;

import com.bennyhuo.annotations.Optional;
import com.sun.tools.javac.code.Symbol;

import javax.lang.model.type.MirroredTypeException;
import javax.lang.model.type.TypeMirror;

/**
 * Created by benny on 1/31/18.
 */

public class OptionalField extends RequiredField {

    private String value;
    private TypeMirror creator;

    public OptionalField(Symbol.VarSymbol symbol) {
        super(symbol, false);
        Optional optional = symbol.getAnnotation(Optional.class);
        value = optional.value();
        try {
            optional.creator();
        } catch (MirroredTypeException e) {
            creator = e.getTypeMirror();
        }
    }

    public String getValue() {
        return value;
    }

    public TypeMirror getCreator() {
        return creator;
    }
}

File: compiler/src/main/java/com/bennyhuo/tieguanyin/compiler/basic/entity/OptionalField.kt
package com.bennyhuo.tieguanyin.compiler.basic.entity

import com.bennyhuo.aptutils.types.isSameTypeWith
import com.bennyhuo.tieguanyin.annotations.Optional
import javax.lang.model.element.VariableElement
import javax.lang.model.type.TypeKind

/**
 * Created by benny on 1/31/18.
 */
class OptionalField(element: VariableElement, optional: Optional) : Field(element) {

    var defaultValue: Any? = null
        private set

    override val prefix = "OPTIONAL_"

    init {
        val fieldType = element.asType()
        defaultValue = when (fieldType.kind) {
            TypeKind.BOOLEAN -> optional.booleanValue
            TypeKind.CHAR -> "'${optional.charValue}'"
            TypeKind.BYTE -> "(byte) ${optional.byteValue}"
            TypeKind.SHORT -> "(short) ${optional.shortValue}"
            TypeKind.INT -> optional.intValue
            TypeKind.LONG -> "${optional.longValue}L"
            TypeKind.FLOAT -> "${optional.floatValue}f"
            TypeKind.DOUBLE -> optional.doubleValue
            else -> if (fieldType.isSameTypeWith(String::class)) {
                //注意字面量的引号
                """"${optional.stringValue}""""
            } else {
                null
            }
        }
    }

    override fun asKotlinTypeName() = super.asKotlinTypeName().asNullable()

    override fun compareTo(other: Field): Int {
        return if (other is OptionalField) {
            super.compareTo(other)
        } else {
            //如果与 RequiredField 比较，Optional 永远排在后面
            1
        }
    }
}
